crates/biome_js_analyze/src/lint/style/use_nullish_coalescing.rs

```diff
@@ -0,0 +1,221 @@
+    JsSyntaxKind, JsWhileStatement, T,
+};
+use biome_rowan::{AstNode, BatchMutationExt};
+use biome_rule_options::use_nullish_coalescing::UseNullishCoalescingOptions;
+
+declare_lint_rule! {
+    /// Enforce using nullish coalescing operator (`??`) instead of logical or (`||`)
+    /// when providing default values.
+    ///
+    /// The `??` operator only checks for `null` and `undefined`, while `||` checks
+    /// for any falsy value including `0`, `''`, and `false`. This can prevent bugs
+    /// where legitimate falsy values are incorrectly treated as missing.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// const value = count || 0;
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// const value = count ?? 0;
+    /// ```
+    ///
+    /// ## Options
+    ///
+    /// The rule provides the option described below.
+    ///
+    /// ```json,options
+    /// {
+    ///     "options": {
+    ///         "ignoreConditionalTests": true
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// ### ignoreConditionalTests
+    ///
+    /// When this option is set to `true`, the rule will not report `||` operators
+    /// used in conditional test positions (if, while, for, ternary test). This is
+    /// useful because in test positions, the distinction between falsy values and
+    /// nullish values often doesn't matter.
+    ///
+    /// When the option is set to `false`, all `||` operators will be reported
+    /// regardless of their position.
+    ///
+    /// Default: `true`
+    ///
+    /// #### Example with `ignoreConditionalTests: false`
+    ///
+    /// ```json,options
+    /// {
+    ///     "options": {
+    ///         "ignoreConditionalTests": false
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// ```js,expect_diagnostic,use_options
+    /// if (foo || bar) {
+    ///     console.log("test position");
+    /// }
+    /// ```
+    ///
+    /// #### Example with `ignoreConditionalTests: true` (default)
+    ///
+    /// ```json,options
+    /// {
+    ///     "options": {
+    ///         "ignoreConditionalTests": true
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// ```js,use_options
+    /// if (foo || bar) {
+    ///     console.log("test position - not reported");
+    /// }
+    /// ```
+    ///
+    /// ```js,expect_diagnostic,use_options
+    /// const value = foo || bar; // Not in test position - still reported
+    /// ```
+    ///
+    pub UseNullishCoalescing {
+        version: "next",
+        name: "useNullishCoalescing",
+        language: "js",
+        sources: &[RuleSource::EslintTypeScript("prefer-nullish-coalescing").same()],
+        recommended: false,
+        severity: Severity::Information,
+        fix_kind: FixKind::Safe,
+    }
+}
+
+impl Rule for UseNullishCoalescing {
+    type Query = biome_analyze::Ast<JsLogicalExpression>;
```

Comment:
You can combine this all into one rule if you use `declare_node_union!` to query multiple node types. Then you can just branch into different functions based on the query.